Iterate by Tweak factor

jswift-stfc opened the pull request:
Added functionality for iterating gudrun_dcs by tweaking the tweak factor iteratively. Also added test for regular runs of gudrun_dcs and runs of gudrun_dcs that iterate by tweaking the tweak factor. Closes #17.

trisyoungs commented on gudrun_classes/gudrun_file.py:
```diff
@@ -1501,21 +1504,47 @@ def write_out(self, overwrite=False):
             )
         return result
 
+    def iterateByTweakFactor(self, n):
```
I would suggest that its probably better to start adding these "useful" functions into their own separate class(es) (here, `TweakFactorIterator` for instance) which take a `GudrunFile` object as a parameter, and keep `GudrunFile` as pure as possible (just read, write, and run). Otherwise it's going to get very big very quickly!

trisyoungs commented on gudrun_classes/gudrun_file.py:
```diff
@@ -1501,21 +1504,47 @@ def write_out(self, overwrite=False):
+
+            # Write out what we currently have,
+            # and run gudrun_dcs on that file.
+            self.write_out()
+            self.dcs(path=self.outpath)
```
It's only two lines, I know, but part of me wants this to be encapsulated in a single method (`process()` for instance?) which writes the file and runs the binary.

trisyoungs commented on gudrun_classes/gudrun_file.py:
```diff
@@ -1501,21 +1504,47 @@ def write_out(self, overwrite=False):
+            self.dcs(path=self.outpath)
+
+            # Iterate through all samples,
+            # updating their tweak factor from the output of gudrun_dcs.
+            for j, sampleBackground in enumerate(self.sampleBackgrounds):
+                for k, sample in enumerate(sampleBackground.samples):
+                    gud = sample.dataFiles.dataFiles[0].replace(
+                                self.instrument.dataFileType,
+                                "gud"
+                            )
+                    gudFile = GudFile(gud)
+                    tweakFactor = float(gudFile.suggestedTweakFactor.strip())
+                    self.sampleBackgrounds[j].samples[k].sampleTweakFactor = (
+                                                tweakFactor
+                    )
```
A likely use-case is that I will only want to adjust tweak factors on a subset of defined samples. In this case, it looks like your loop would adjust the tweak factors of all samples regardless.  Of course, the old GudrunGUI was limited in the sense that it didn't bother objectifying the data it read in, so when it came to things like samples being enabled / disabled it's all contained in the GUI.  You don't have that problem, however, so you can just add a data member to your sample class that encapsulates this kind of information, but doesn't get written out.

jswift-stfc replied:
That sounds like a suitable enhancement. I will add an attribute such as 'adjustTweakFactor', and add it to the Sample class.

trisyoungs replied:
I was thinking more along the lines of a generic "runThisSample" data member.  An "adjustTweakFactor" is too specific - the method handling the running, reading, and setting of the tweak factors is the only place where that knowledge is important.

trisyoungs commented:
Some architectural suggestions to consider, thinking about the bigger picture.

jswift-stfc commented:
Added process method to GudrunFile class, and removed iterateByTweakFactor method. (closes #20)
Added runThisSample attribute to sample class. (closes #19)
Created TweakFactorIterator class. (closes #21)

trisyoungs reviewed (approved):
:+1:

And I have to say that it is fantastic to finally see the ability to manipulate Gudrun input files in a programmatic way.

jswift-stfc commented:
> programmatic

I agree! Getting to the exciting stuff! :smile:

jswift-stfc commented:
Just waiting for the pipeline to run, and I'll merge if everything is okay there.
